### PR TITLE
[Feature](bangc-ops): add new op indices_convolution_backward_data

### PR DESCRIPTION
**Pull request by duzekunKTH**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new API called indice_convolution_backward_data

## 2. Modification

create bangc-ops/kernels/indice_convolution_backward_data/
create bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/indice_convolution_backward_data/
create docs/bangc-docs/design_docs/indice_convolution_backward_data/indice_convolution_backward_data.md
modify bangc-ops/mlu_op.h

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 1e-5 (float)
- [x] diff2: diff2 <= 1e-5 (float)
- [x] diff1: diff1 <= 3e-3 (half)
- [x] diff2: diff2 <= 3e-3 (half)

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1 <br> U4       |
|        3       |Layouts                               |          filters: NHWC 、NCHW、HWCN、ARRAY、NCDHW、NDHWC;<br>other tensors: ARRAY      |
|        4       |Whether multi-dimensions are supported|   filters tensor supports 4D or 5D<br>other tensors: fix dimensions  |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |          filters: half / float            |
|        7       |Whether there is size limit           |  could not reach 2G num |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |

```C++
[^      OK ] /projs/platform/duzekun/dzk_mluops/indice_convbpdata_case/generator/indice_convolution_backward_data/indice_convolution_backward_data_data_included_int32_float16_1675147683889.pb
[       OK ] indice_convolution_backward_data/TestSuite.mluOp/117 (344 ms)
[----------] 118 tests from indice_convolution_backward_data/TestSuite (3052 ms total)

[----------] Global test environment tear-down
[ SUMMARY  ] Total 118 cases of 1 op(s).
ALL PASSED.
[==========] 118 test cases from 1 test suite ran. (3437 ms total)
[  PASSED  ] 118 test cases.
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|op_name|    |    |     |    |    |    |     |
|op_name|    |    |     |    |    |    |     |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|    |    |    |    |    |    |     |
|op_name|    |    |    |    |    |    |     |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/indice_convolution_backward_data/indice_convolution_backward_data.cpp [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <vector>
+
+#include "test/mlu_op_gtest/pb_gtest/src/zoo/indice_convolution_backward_data/indice_convolution_backward_data.h"
```
Comment (WangYuan-cn) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/indice_convolution_backward_data/indice_convolution_backward_data.cpp:25]:
```suggestion
#include "indice_convolution_backward_data.h"
```

File: bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+inline int getMaxNumInArray(const int64_t arr[], const int num) {
+  int max_num = (int)(arr[0]);
+  for (int i = 1; i < num; ++i) {
+    max_num = max_num > (int)(arr[i]) ? max_num : (int)(arr[i]);
+  }
+  return max_num;
+}
+
+inline int getSizeOfDataType(const mluOpDataType_t dtype) {
```
Comment (defei-coder) [inline bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.h:37]: bangc-ops/core/type.cpp中有相同功能的函数

File: bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+  PARAM_CHECK(api, input_grad_desc->dtype == MLUOP_DTYPE_FLOAT ||
+                       input_grad_desc->dtype == MLUOP_DTYPE_HALF);
+  PARAM_CHECK(api,
+              indice_pairs_desc->dtype == MLUOP_DTYPE_INT32);  // check layout
+  bool layout_check = filters_desc->layout == MLUOP_LAYOUT_NHWC ||
+                      filters_desc->layout == MLUOP_LAYOUT_NCHW ||
+                      filters_desc->layout == MLUOP_LAYOUT_HWCN ||
+                      filters_desc->layout == MLUOP_LAYOUT_NCDHW ||
+                      filters_desc->layout == MLUOP_LAYOUT_NDHWC ||
+                      filters_desc->layout == MLUOP_LAYOUT_ARRAY;
```
[the remark this answers is not in the document]
Comment (duzekunKTH) [reply bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp:65]: array 指代 DHWCN，除了这些layout，库里还定义了其他奇怪的layout，防那些内容。

File: bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+                                               &transpose_workspace_size_));
+    MLUOP_CHECK(mluOpDestroyTransposeDescriptor(trans_desc));
+    transpose_workspace_size = (uint64_t)transpose_workspace_size_;
+  }
+  output_grad_condence_size = max_indice_num * output_grad_desc->dims[1] *
+                              getSizeOfDataType(filters_desc->dtype);
+  input_grad_condence_size =
+      max_indice_num * input_grad_desc->dims[1] *
+      getSizeOfDataType(filters_desc->dtype);  // matmul workspace
+  {
```
[the remark this answers is not in the document]
Comment (duzekunKTH) [reply bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp:344]: 仅隔离代码方便阅读，也可以提前释放变量

File: docs/bangc-docs/design_docs/indice_convolution_backward_data/indice_convolution_backward_data.md [hunk 5]
```diff
@@ -0,0 +1,402 @@
+# IndiceConvolutionBackwardData 算子开发设计方案
+* #### 文档基本信息
+
+| 算子名称    | IndiceConvolutionBackwardData                             |
+| ----------- | ------------------------------------------------------------|
+| 编制人/日期 | 杜泽坤/2022-11-28                                            |
+| 审批人/日期 | XX/2020-11-2                                               |
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/indice_convolution_backward_data/indice_convolution_backward_data.md:7]: 添加下reviewers & chengwei

File: docs/bangc-docs/design_docs/indice_convolution_backward_data/indice_convolution_backward_data.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+                                   const size_t workspace_size,
+                                   const mluOpTensorDescriptor_t input_grad_desc,
+                                   void *input_grad);
+```
+
+## 3 实现方案设计
+
+### 3.1 实现方案
+
+workspace 空间分成两份，分别取名为 filter_transpose、output_grad_condence、input_grad_condence、workspace_input_grad_tmp、workspace_addn。
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/indice_convolution_backward_data/indice_convolution_backward_data.md:171]: workspace 空间份数需要确认

**Review by defei-coder — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * The size of extra workspace is based on the given information of the indice
+ * convolution backward data operation, including the input descriptor
+ * \b input_grad_desc, the filter descriptor \b filter_desc, the indice pairs
+ * descriptor \b indice_pairs_desc, the output descriptor \b indice_pairs_desc, 
+ * the array \b indice_num and the scaler \b inverse. For more information
+ * about the workspace, see "Cambricon BANGC OPS User Guide".
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * the IndiceConvolutionBackwardData operation. For detailed information, see
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7449]: ::IndiceConvolutionBackwardData

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @param[in] indice_pairs_desc
+ * The descriptor of the indice_pairs tensor. For detailed information, see
+ * ::mluOpTensorDescriptor_t.
+ * @param[in] input_grad_desc
+ * The descriptor of the input_grad tensor. For detailed information, see
+ * ::mluOpTensorDescriptor_t.
+ * @param[in] indice_num
+ * The array to describe the valid data number in the indice_pairs tensor.
+ * @param[in] inverse
+ * The param to describe the operation performs deconvolution logic or not.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7466]: The parameter to describe whether the operation performs deconvolution logic.

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that is used in
+ * the IndiceConvolutionBackwardData operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7469]: ::IndiceConvolutionBackwardData

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par API Dependency
+ * - This function must be called before the ::mluOpIndiceConvolutionBackwardData
+ *   function.
+ * - The ::mluOpCreateTensorDescriptor and ::mluOpSetTensorDescriptor functions
+ *   to create and set the tensor descriptor \b output_grad_desc, \b filters_desc,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7477]: 删除to

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *   \b indice_pairs_desc and \b input_grad_desc before calling this function.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7478]: before this function is called.

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+                                                   const mluOpTensorDescriptor_t filters_desc,
+                                                   const mluOpTensorDescriptor_t indice_pairs_desc,
+                                                   const mluOpTensorDescriptor_t input_grad_desc,
+                                                   const int64_t indice_num[],
+                                                   const int64_t inverse,
+                                                   size_t *workspace_size);
+
+// Group:IndiceConvolutionBackwardData
+/*!
+ * @brief Performs the backpropagation of a indice convolution operation to
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7501]: a > an

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * compute the gradient of input \b input_grad based on the gradient of response
+ * \b output_grad, the filter tensor \b filter, the indice tensor \b indice_pairs
+ * and helper parameters: array \b indice_num, scaler \b inverse and \b sub_m.
+ *
+ * The tensors \b input_grad and \b output_grad is reordered from origin input
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7506]: is > are

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * gradient and output gradient. The tensor \b indice_pairs describes the
+ * calculation logic between the tensors \b input_grad and \b output_grad.
+ * Every pair in \b indice_pairs points to part of \b filters and the single
+ * line in \b input_grad and \b output_grad. Every single line in \b input_grad
+ * is calculated from matmul calculation logic with data mensioned above.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7511]: mensioned拼写错误  mentioned

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * the IndiceConvolutionBackwardData operation. For detailed information, see
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7515]: ::IndiceConvolutionBackwardData

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @param[in] indice_pairs_desc
+ * The descriptor of input data \b indice_pairs, which contains dimension,
+ * data type and data layout.
+ * @param[in] indice_pairs
+ * Pointer to the MLU memory that stores the indice_pairs tensor. It is used to
+ * specify the calculation pairs between \b input_grad and \b output_grad.
+ * @param[in] indice_num
+ * The array to describe the valid data number in \b indice_pairs.
+ * @param[in] inverse
+ * The param to describe the operation performs deconvolution logic or not.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7538]: The parameter to describe whether the operation performs deconvolution logic or not.

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @param[in] sub_m
+ * The param to describe the operation performs sub_m convolution logic or
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7540]: The parameter to describe whether the operation performs sub_m convolution logic or

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * sparce convolution logic.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the
+ * IndiceConvolutionBackwardData operation. For more information about workspace,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7544]: ::IndiceConvolutionBackwardData

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * see "Cambricon BANGC OPS User Guide".
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that is used in
+ * the IndiceConvolutionBackwardData operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7548]: ::IndiceConvolutionBackwardData

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *   - filters tensor: MLUOP_LAYOUT_ARRAY, MLUOP_LAYOUT_HWCN, MLUOP_LAYOUT_NCHW,
+ *     MLUOP_LAYOUT_NHWC, MLUOP_LAYOUT_NCDHW, MLUOP_LAYOUT_NDHWC
+ *   - input_grad tensor: MLUOP_LAYOUT_ARRAY
+ *
+ * @par Scale Limitation
+ * - The \b output_grad and \b input_grad are 2D arrays.
+ * - The \b indice_pairs is 3D array.
+ * - The \b filter is 4D or 5D tensor.
+ * - The dims[1] of \b indice_pairs should be equal to 2.
+ * - When \b filter is a 4D tensor, the dims[0] of \b indice_pairs shoule be
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7588]: shoule   拼写错误

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *   equal to H * W corresponding to filter layout.
+ * - When \b filter is a 5D tensor, the dims[0] of \b indice_pairs shoule be
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7590]: shoule 拼写错误

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -7432,6 +7432,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * - The dims[1] of \b input_grad should be equal to C corresponding to filter layout.
+ * - The dims[0] of \b input_grad should be equal to the dims[2] of \b indice_pairs.
+ * - Each value in array \b indice_num should be no smaller than 0, no
+ *   larger than the dims[0] of \b output_grad and no larger than the dims[2]
+ *   of \b indice_pairs. 
+ * - The value of \b inverse and \b sub_m should be 0 or 1.
+ *
+ * @par API Dependency
+ * - The function ::mluOpGetIndiceConvolutionBackwardDataWorkspaceSize should
+ *   be called to get the extra space size before calling this function.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7602]: before calling this function > before this function is called.

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments and refer to the guide wiki.

**Review by Unireverse — APPROVED** [review]

**Review by GraceGuanH — APPROVED** [review]

**Review by WangYuan-cn — APPROVED** [review]

File: bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp [hunk 23]
```diff
@@ -0,0 +1,716 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "kernels/indice_convolution_backward_data/indice_convolution_backward_data.h"
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/indice_convolution_backward_data/indice_convolution_backward_data.cpp:23]: **#include "indice_convolution_backward_data.h"**

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -7432,6 +7435,203 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * ::mluOpTensorDescriptor_t.
+ * @param[in] indice_num
+ * The array to describe the valid data number in the indice_pairs tensor.
+ * @param[in] inverse
+ * The parameter to describe whether the operation performs deconvolution logic.
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that is used in the
+ * ::mluOpIndiceConvolutionBackwardData operation.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:7474]: 缺少 MLUOP_STATUS_ARCH_MISMATCH

**Pull request merged**